tests/test/subtype.rs

```diff
@@ -0,0 +1,314 @@
+use super::*;
+
+#[test]
+fn variance_lowering() {
+    lowering_success! {
+        program {
+            #[variance(Invariant, Covariant)]
+            struct Foo<T, U> {}
+            struct Bar<T, U> {}
+            #[variance(Invariant, Contravariant)]
+            fn foo<T, U>(t: T, u: U);
+            fn bar<T, U>(t: T, u: U);
+        }
+    }
+}
+
+#[test]
+fn subtype_simple() {
+    test! {
+        program {
+            struct Foo { }
+        }
+
+        goal {
+            Subtype(Foo, Foo)
+        } yields {
+            "Unique"
+        }
+    }
+}
+
+/// Test that `Foo<'a>` and `Foo<'b>` can be subtypes
+/// if we constrain the lifetimes `'a` and `'b` appropriately.
+#[test]
+fn struct_lifetime_variance() {
+    test! {
+        program {
+            #[variance(Covariant)]
+            struct Foo<'a> { }
+        }
+
+        goal {
+            forall<'a, 'b> {
+                Subtype(Foo<'a>, Foo<'b>)
+            }
+        } yields {
+            "Unique; substitution [], lifetime constraints [\
+            InEnvironment { environment: Env([]), goal: '!1_1: '!1_0 } \
+            ]"
+        }
+    }
+}
+
+/// Test that `&'a u32 <: &'b u32` if `'a: 'b`
+#[test]
+fn ref_lifetime_variance() {
+    test! {
+        program {
+        }
+
+        goal {
+            forall<'a, 'b> {
+                Subtype(&'a u32, &'b u32)
+            }
+        } yields {
+            // Seems good!
+            "Unique; substitution [], lifetime constraints [\
+            InEnvironment { environment: Env([]), goal: '!1_0: '!1_1 }\
+            ]"
+        }
+    }
+}
+
+#[test]
+fn fn_lifetime_variance_args() {
+    test! {
+        program {
+        }
+
+        goal {
+            for<'a, 'b> fn(&'a u32, &'b u32) = for<'a> fn(&'a u32, &'a u32)
+        } yields {
+            "Unique;substitution [], lifetime constraints [\
+            InEnvironment { environment: Env([]), goal: '!1_0: '!1_1 }, \
+            InEnvironment { environment: Env([]), goal: '!1_1: '!1_0 }]"
```

Comment:
This doesn't look right, these constraints are not solvable.

Reply:
But these are? This just says that `'a == 'b`

Reply:
These are placeholders, so relating them is an error. The requirements here should be `!1_0: '?0`, `!1_1: ?0`, `'!2_0: '?1`, `'!2_0: '?2` (the same as subtyping in both directions).

Reply:
But we have allowed unifying placeholders? (On master, [here](https://github.com/rust-lang/chalk/blob/65daa08f07fcea1314c0bffd731e1fef1b76d619/chalk-solve/src/infer/unify.rs#L385) is unification, [here](https://github.com/rust-lang/chalk/blob/65daa08f07fcea1314c0bffd731e1fef1b76d619/tests/test/existential_types.rs#L310) an example where two lifetime placeholders are equated).

Reply:
yes, but region inference will error on such constraints.